### ISSUE TITLE
Feature request: unused-annotation rule

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4290]
Explicit type annotations for local variables are often used only to steer the type checker and e.g. remove a false positive error or correct for imprecise annotations elsewhere. But if the reason for the annotation is later removed, for example by adding the missing annotation or fixing an inference bug in the type checker, the annotation stays around and there's nothing telling you to remove it. However, I would prefer to remove such annotations in order to keep the code simpler and more concise. This is similar to the rules against unused casts and type ignores, but a bit more opinionated as users may legitimately want to keep annotations around for various reasons, so it should likely stay disabled by default.

Example code on which this should trigger:

```python
def f() -> int:
    return 3

def g() -> None:
    x: int = f() # E: unused-annotation
    print(x)
```

**Comment by carljm** [comment]
In your example, the `x: int` declaration defines an upper bound on the type of `x` and would cause a later non- `int` assignment to `x` to fail. Without it, a later non `int` assignment would succeed. So the `x: int` declaration is not meaningless in this example (if we consider the context of potential future changes to the code).

But I guess the documentation for this opt-in diagnostic should just mention this. Effectively, opting in to this diagnostic would be saying "I don't care about the use of declarations to restrict later assignments" -- or "if I want the upper-bound declared-type effect, I'll use a declaration with no assigned value" (I assume the rule would not fire for those?)

I don't think any other type checker currently offers this, so I don't think it would an in-scope feature for stable release.
